client/src/redux/searchReducer.ts

```diff
@@ -0,0 +1,53 @@
+import { SEARCH_LOCALSTORAGE_LABEL } from 'constants/labels';
+import { AnyAction } from 'redux';
+import { SearchState } from 'types';
+import { getFromStorage, setInStorage } from '../API/localStorage';
+
+export const SEARCH_SAVE_CURRENT_LOCATION = 'SAVE_CURRENT_LOCATION';
+export const SEARCH_SAVE_LOCATION = 'SAVE_LOCATION_SEARCH';
+
+const initialState: SearchState = {
+  prevSearches: getFromStorage(SEARCH_LOCALSTORAGE_LABEL) || []
+};
+
+export const searchReducer = (
+  state: SearchState = initialState,
+  action: AnyAction
+): SearchState => {
+  if (action.type === SEARCH_SAVE_LOCATION) {
+    if (action.payload === '') {
+      return state;
+    }
+    if (state.prevSearches.includes(action.payload)) {
+      const filteredPrevSearches = [...state.prevSearches].filter(element => {
+        return element !== action.payload;
+      });
+      const prevSearches = [action.payload, ...filteredPrevSearches];
+      setInStorage(SEARCH_LOCALSTORAGE_LABEL, prevSearches);
+      return {
+        ...state,
+        prevSearches
+      };
+    }
+    if (state.prevSearches.length > 3) {
+      const slicedPrevSearches = [...state.prevSearches].slice(0, 3);
+      const prevSearches = [action.payload, ...slicedPrevSearches];
+      setInStorage(SEARCH_LOCALSTORAGE_LABEL, prevSearches);
+      return {
+        ...state,
+        prevSearches
+      };
+    }
+    const prevSearches = [action.payload, ...state.prevSearches];
+    setInStorage(SEARCH_LOCALSTORAGE_LABEL, prevSearches);
+    return {
+      ...state,
+      prevSearches
+    };
+  }
+  return state;
+};
```

Comment:
1. reducer should be a pure function, so it's not recommended to add functions related to local storage here. you can use subscribe method to update local storage values
https://stackoverflow.com/a/37690899
2. i see 3 times almost the same code lines, that means the code should be refactored:
```
 const prevSearches = [action.payload, ...slicedPrevSearches];
setInStorage(SEARCH_LOCALSTORAGE_LABEL, prevSearches);
    return {
      ...state,
      prevSearches
    };

```

Reply:
I solved this problem with action creator as I have already did before with feedback reducer. I will definitely read and learn about subscribe method in the near future. It is too late and not enough time for my tired head :)